### ISSUE TITLE
Write terraform + ansible scripts to bring up the elasticsearch + graylog cluster for logging.

### ISSUE THREAD

**Issue by gidoBOSSftw5731** [RIT-GCI/code-red#6]

**Comment by gidoBOSSftw5731** [comment]
Ansible is running and the code is mostly in place to run playbooks, just needs to be finished and tested. Need to figure out a strategy to load and run new playbooks without restarting ansible.

**Comment by gidoBOSSftw5731** [comment]
@ashim-mahara Elastic search is now initialized with a default config, let me know if you want it configured before a merge

**Comment by gidoBOSSftw5731** [comment]
Todo: Add slave nodes and automatically attach them to the master. Use self-signed SSL between them.

**Comment by ashim-mahara** [comment]
> Todo: Add slave nodes and automatically attach them to the master. Use self-signed SSL between them.

Heap Memory: 30 GB, Total Instance Memory 64 GB. Linux cache will be used too.

**Comment by gidoBOSSftw5731** [comment]
> > Todo: Add slave nodes and automatically attach them to the master. Use self-signed SSL between them.
> 
> Heap Memory: 30 GB, Total Instance Memory 64 GB. Linux cache will be used too.

That's quite a lot of nodes, is there a particular reason we need so much? I'm thinking for now we start around 4-8G of memory and then we can always add more if we need them.

**Comment by ashim-mahara** [comment]
> That's quite a lot of nodes, is there a particular reason we need so much? I'm thinking for now we start around 4-8G of memory and then we can always add more if we need them.

Elasticsearch is notoriously memory-hungry. You'll see performance degradation for 8-gig setups. This has to do with how Elasticsearch stores the indexes. The most used indices/shards are stored in the memory for faster access. Each index will have a shard (indices are split into shards) with a maximum of 20 gigs of documents. I am not sure how many logs we will be ingesting per day but this is good for 100 million events per day. You can even configure hot and cold nodes but I don't think we would need that here.

It's cool if you want to set it to 4/8 for development purposes but if we're constantly pushing logs then it will be unusable after a certain amount of logs have been stored. Possibly in the 50 gigs range. I remember Forrest talking about storing previous engagement data too, so storing that will require more storage and thus more heap. And if accessing it via Kibana/Grafana, then the retrieval process will require even more.

Optimal configuration for Elasticsearch requires building schema, designing the hot/cold, configuring shard sizes, and shard allocation/routing among others. These require a lot of effort to configure and test the results and probably can be dwarfed by throwing more resources at the cluster. We have 1 TB of RAM available for the project. 
:man_shrugging: 

TLDR; elastic search is memory hungry, optimal configuration is time-consuming, and do the brute approach if you've got the resources.

**Comment by gidoBOSSftw5731** [comment]
Elastic search now deploys, I'll test it later to make sure it's actually making a cluster later and then configure it with graylog

**Comment by gidoBOSSftw5731** [comment]
Elastic does NOT work security enabled at the moment, complains about no authentication schemes available. It does work without security enabled, but it's not supported to run encryption on only HTTP.

```
[2023-11-06T22:12:48,503][WARN ][o.e.t.TcpTransport       ] [elasticmaster] exception caught on transport layer [Netty4TcpChannel{localAddress=/10.10.40.201:9300, remoteAddress=/10.10.40.202:56130, profile=default}], closing connection
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: No available authentication scheme
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:477) ~[netty-codec-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) ~[netty-codec-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:620) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:583) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [netty-common-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.66.Final.jar:4.1.66.Final]
	at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: javax.net.ssl.SSLHandshakeException: No available authentication scheme
	at sun.security.ssl.Alert.createSSLException(Alert.java:131) ~[?:?]
	at sun.security.ssl.Alert.createSSLException(Alert.java:117) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:357) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:313) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:304) ~[?:?]
	at sun.security.ssl.CertificateMessage$T13CertificateProducer.onProduceCertificate(CertificateMessage.java:972) ~[?:?]
	at sun.security.ssl.CertificateMessage$T13CertificateProducer.produce(CertificateMessage.java:961) ~[?:?]
	at sun.security.ssl.SSLHandshake.produce(SSLHandshake.java:440) ~[?:?]
	at sun.security.ssl.ClientHello$T13ClientHelloConsumer.goServerHello(ClientHello.java:1246) ~[?:?]
	at sun.security.ssl.ClientHello$T13ClientHelloConsumer.consume(ClientHello.java:1182) ~[?:?]
	at sun.security.ssl.ClientHello$ClientHelloConsumer.onClientHello(ClientHello.java:840) ~[?:?]
	at sun.security.ssl.ClientHello$ClientHelloConsumer.consume(ClientHello.java:801) ~[?:?]
	at sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:396) ~[?:?]
	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:480) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1277) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1264) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:712) ~[?:?]
	at sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1209) ~[?:?]
	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1550) ~[netty-handler-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1396) ~[netty-handler-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1237) ~[netty-handler-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1286) ~[netty-handler-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507) ~[netty-codec-4.1.66.Final.jar:4.1.66.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446) ~[netty-codec-4.1.66.Final.jar:4.1.66.Final]
	... 16 more
```

@ashim-mahara If you have any input there, that'd be appreciated, otherwise I'll keep kicking at this until it cooperates.

**Comment by ashim-mahara** [comment]
> Elastic does NOT work security enabled at the moment, complains about no authentication schemes available. It does work without security enabled, but it's not supported to run encryption on only HTTP.
> 
> ```
> [2023-11-06T22:12:48,503][WARN ][o.e.t.TcpTransport       ] [elasticmaster] exception caught on transport layer [Netty4TcpChannel{localAddress=/10.10.40.201:9300, remoteAddress=/10.10.40.202:56130, profile=default}], closing connection
> io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: No available authentication scheme
> 	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:477) ~[netty-codec-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) ~[netty-codec-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:620) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:583) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) [netty-common-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.66.Final.jar:4.1.66.Final]
> 	at java.lang.Thread.run(Thread.java:833) [?:?]
> Caused by: javax.net.ssl.SSLHandshakeException: No available authentication scheme
> 	at sun.security.ssl.Alert.createSSLException(Alert.java:131) ~[?:?]
> 	at sun.security.ssl.Alert.createSSLException(Alert.java:117) ~[?:?]
> 	at sun.security.ssl.TransportContext.fatal(TransportContext.java:357) ~[?:?]
> 	at sun.security.ssl.TransportContext.fatal(TransportContext.java:313) ~[?:?]
> 	at sun.security.ssl.TransportContext.fatal(TransportContext.java:304) ~[?:?]
> 	at sun.security.ssl.CertificateMessage$T13CertificateProducer.onProduceCertificate(CertificateMessage.java:972) ~[?:?]
> 	at sun.security.ssl.CertificateMessage$T13CertificateProducer.produce(CertificateMessage.java:961) ~[?:?]
> 	at sun.security.ssl.SSLHandshake.produce(SSLHandshake.java:440) ~[?:?]
> 	at sun.security.ssl.ClientHello$T13ClientHelloConsumer.goServerHello(ClientHello.java:1246) ~[?:?]
> 	at sun.security.ssl.ClientHello$T13ClientHelloConsumer.consume(ClientHello.java:1182) ~[?:?]
> 	at sun.security.ssl.ClientHello$ClientHelloConsumer.onClientHello(ClientHello.java:840) ~[?:?]
> 	at sun.security.ssl.ClientHello$ClientHelloConsumer.consume(ClientHello.java:801) ~[?:?]
> 	at sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:396) ~[?:?]
> 	at sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:480) ~[?:?]
> 	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1277) ~[?:?]
> 	at sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1264) ~[?:?]
> 	at java.security.AccessController.doPrivileged(AccessController.java:712) ~[?:?]
> 	at sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1209) ~[?:?]
> 	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1550) ~[netty-handler-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1396) ~[netty-handler-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1237) ~[netty-handler-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1286) ~[netty-handler-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507) ~[netty-codec-4.1.66.Final.jar:4.1.66.Final]
> 	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446) ~[netty-codec-4.1.66.Final.jar:4.1.66.Final]
> 	... 16 more
> ```
> 
> @ashim-mahara If you have any input there, that'd be appreciated, otherwise I'll keep kicking at this until it cooperates.

This is possibly due to the Certificate Authority not being provided to the Elasticsearch. Create a CA then use that CA to sign other certificates. You also have to provide the CA to the ES Security plugin. I am going off the top, and the issue might be entirely different but this is a major issue most of the time.

**Comment by ashim-mahara** [comment]
partly fixed with #17
